userdocs/diagnostics/associated-type-requirements.md

```diff
@@ -0,0 +1,23 @@
+Using Protocols with `Self` or Associated Type Requirements
+---
+Protocols in Swift may be used as types, as part of a generic constraint, or as part of an opaque result type.
+```
+// CustomStringConvertible can be used as a type.
+func foo(bar: CustomStringConvertible) { /* ... */ }
+
+// ...or as a generic constraint on 'T'.
+func bar<T: CustomStringConvertible>(baz: T) { /* ... */ }
+
+// ...or as part of an opaque result type.
+func baz() -> some CustomStringConvertible { /* ... */ }
+```
+
+While all Swift protocols can be used as generic constraints and as part of opaque result types, not all protocols can be used as types in general. Specifically, if a protocol has a requirement which references `Self` or an associated type, it cannot be used as a type. One such protocol is `Identifiable`, which has the requirement `var id: ID { get }`, where `ID` is an associated type. As a result, the following code is not allowed:
+```
+func foo(bar: Identifiable) { /* ... */ }
+// error: protocol 'Identifiable' can only be used as a generic constraint because it has Self or associated type requirements
+```
+
+Protocols like `Identifiable` which have `Self` or associated type requirements cannot be used as types because the type named `Identifiable` does not conform to the protocol named `Identifiable`. It is unable to satisfy the `var id: ID { get }` requirement because no associated type is specified.
```

Comment:
Is this necessarily true? Cherry picking few quotes from Joe: 

> The limitations are because of missing implementation details and incomplete language design rather than any fundamental difference. Where there are complications of the sort Dave raises, where an existential has a different set of behavior or methods from conforming types, that doesn’t even really arise because of associated types, really, but because of covariance and contravariance. It’s been an unfortunate byproduct of the “self or associated type” restriction that it’s introduced this confusion.

>  passing an existential as an argument is isomorphic to passing a generic argument with its own type variable—anything that's possible with one representation is possible on the other. We've even discussed giving them the same underlying ABI.

> Using an existential does not require anything about the existential type conforming to the protocol.

> that's not because of associated types specifically but because of contravariant self/associated type requirements. If a protocol requirement takes a value of Self type or of an associated type as an argument, then there isn't any automatic way to generalize from the implementations for individual types to an implementation that works dynamically with heterogeneous types.

Reply:
I suppose this is fine for now, but I don't know how accurate and detailed we want these notes to be.

Reply:
I think that it doesn't have to be perfect, we can definitely iterate here...

Reply:
I agree what's written here isn't 100% correct, but I think a more complete explanation would be pretty confusing most of the times this error is encountered. I can definitely continue to iterate on this if I come up with a better way to explain it

Reply:
Wouldn't `Equatable` be sufficient as a widely used example? It contains `(Self, Self) -> Self` requirement for which there isn't an automatic way to generalize from implementations for individual types to an implementation that works dynamically with heterogeneous types.

Reply:
Maybe there's a way to rephrase that sentence to focus a little less on the lack of self-conformance, I'll hold off on merging this a bit longer to think it over. I do like using `Identifiable` as an example because the `id` requirement is very simple and easy to understand

Reply:
I had some time to take another look at this today. I think @xwu's comment is really helpful in that it explains a major reason the protocol can't be used this way is that it wouldn't be very useful (and often would be the wrong solution) if it was allowed.

I've updated the paragraph from

"Protocols like Identifiable which have Self or associated type requirements cannot be used as types because the type named Identifiable does not conform to the protocol named Identifiable. It is unable to satisfy the var id: ID { get } requirement because no associated type is specified."

to 

"Protocols like `Identifiable` which have `Self` or associated type requirements cannot be used as types because such types would rarely be useful in practice. They would be unable to allow use of `Self` or associated type requirements like `var id: ID { get }` because the associated type is not specified."